Overlay overhaul and tooltip refactor

jherico opened the pull request:
In order to prevent the overlay rendering messing up the GL state, I've moved it to after the primary rendering.  The result is that there is a one frame latency on the content of the overlay, but no risk of breaking the GL state for the main scene rendering.

hifi-gustavo commented:
Build console output for commit: e44b21ee0bef85200a1d1ed48cd2d24463be9023 available at: https://gist.github.com/9e3764e8f52d18446ef6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg

ZappoMan commented:
This build still breaks the default skybox/atmosphere and stars... it's rendering just all white for them. So I don't think we want to merge this till that issue is fixed.

hifi-gustavo commented:
Build console output for commit: 07845df4dc5ea1b9145146b00d889266eb4e7fc5 available at: https://gist.github.com/9e3764e8f52d18446ef6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg

hifi-gustavo commented:
Build console output for commit: 02d33f6086f98092c5168401597740f2ba549e77 available at: https://gist.github.com/9e3764e8f52d18446ef6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg

hifi-gustavo commented:
Build console output for commit: 4762e1a00c9ac54bf395365791d4eb068653d53e available at: https://gist.github.com/9e3764e8f52d18446ef6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg

hifi-gustavo commented:
Build console output for commit: 2bf53b625e4b2cc7f6e210aae456cac633fa0a77 available at: https://gist.github.com/9e3764e8f52d18446ef6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg

hifi-gustavo commented:
Build console output for commit: 5ded9e7eb90886fa1a21759027d80601a4d39db8 available at: https://gist.github.com/9e3764e8f52d18446ef6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg

ZappoMan commented:
Mac desktop stats/overlay display is borked... it's not visible at all. Also in HMD and 3D TV mode -- nothing displayed.

hifi-gustavo commented:
Build console output for commit: 1e90aaf9c36a197c076e328f0f896203bd66b081 available at: https://gist.github.com/9e3764e8f52d18446ef6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg

hifi-gustavo commented:
Build console output for commit: 4282c044ef4d4e0b82b84a33c0391003214f214a available at: https://gist.github.com/9e3764e8f52d18446ef6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg

hifi-gustavo commented:
Build console output for commit: 11fff7c6e503866d803e78c917ffb53afe1351f2 available at: https://gist.github.com/189204583104c98d3003<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg

hifi-gustavo commented:
Build console output for commit: ae1c9603b9d5453050fe74bd469c42a69ad12ee6 available at: https://gist.github.com/9e3764e8f52d18446ef6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg

hifi-gustavo commented:
Build console output for commit: c8cc3a46b37e2fc1e0c493a88da5e0ee5fe860eb available at: https://gist.github.com/9e3764e8f52d18446ef6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg

hifi-gustavo commented:
Build console output for commit: 8a1b51907138d8c09cac0f0182c20dc4c81b692a available at: https://gist.github.com/9e3764e8f52d18446ef6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg

hifi-gustavo commented:
Build console output for commit: 1705eb8f80bfdc5902d4c353552307ecd44b9a49 available at: https://gist.github.com/9e3764e8f52d18446ef6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg

hifi-gustavo commented:
Build console output for commit: 5c4fa77acdaa56f8e66482b7ba9d5da7a72b26e7 available at: https://gist.github.com/9e3764e8f52d18446ef6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg

ZappoMan commented:
test this please

hifi-gustavo commented:
Build console output for commit: 5c4fa77acdaa56f8e66482b7ba9d5da7a72b26e7 available at: https://gist.github.com/9e3764e8f52d18446ef6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg

hifi-gustavo commented:
Build console output for commit: 440491700303d67a925bc48707c7cd7e926f7fb1 available at: https://gist.github.com/9e3764e8f52d18446ef6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg

hifi-gustavo commented:
Build console output for commit: 0cdee2d6ec9c092db92bddd9ae20a777e2c3ed8a available at: https://gist.github.com/9e3764e8f52d18446ef6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg

hifi-gustavo commented:
Build console output for commit: 1243c547671a9fa1ae5664fa1b75ffb56dc7a424 available at: https://gist.github.com/9e3764e8f52d18446ef6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg

hifi-gustavo commented:
Build console output for commit: e75b1a6794350b675cd62a6a36466711036de5c4 available at: https://gist.github.com/9e3764e8f52d18446ef6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg

hifi-gustavo commented:
Build console output for commit: 68595ef5dc90fb2eab487ef050677c0195109f6d available at: https://gist.github.com/9e3764e8f52d18446ef6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg

hifi-gustavo commented:
Build console output for commit: edfb25bfe2290f4f1fbe66ec9b549bf7deaa891c available at: https://gist.github.com/9e3764e8f52d18446ef6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg

hifi-gustavo commented:
Build console output for commit: 48b27acc79755074ca00317ce6f04abfb510e74e available at: https://gist.github.com/9e3764e8f52d18446ef6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg

samcake commented on interface/src/ui/ApplicationOverlay.cpp:
```diff
@@ -185,680 +90,75 @@ ApplicationOverlay::~ApplicationOverlay() {
+        gpu::Batch batch;
+        auto geometryCache = DependencyManager::get<GeometryCache>();
+        geometryCache->useSimpleDrawPipeline(batch);
+        batch._glDisable(GL_DEPTH_TEST);
```
Since now you are using the PIpeline, all the state of all the funcs is assigned correclty.

SO if you want a different state, you need a different Pipeline with a State that is different.

The gpu::State by default is defined in the header gpu/State.h, 

Everything about DepthTest is expressed in a struct called  DepthTest
It's off by default. 

So please can you remove these  gl calls which are not needed anymore

jherico replied:
fixed

samcake commented on interface/src/ui/ApplicationOverlay.cpp:
```diff
@@ -185,680 +90,75 @@ ApplicationOverlay::~ApplicationOverlay() {
+        auto geometryCache = DependencyManager::get<GeometryCache>();
+        geometryCache->useSimpleDrawPipeline(batch);
+        batch._glDisable(GL_DEPTH_TEST);
+        batch._glDisable(GL_LIGHTING);
```
Not needed anymore

jherico replied:
fixed

samcake commented on interface/src/ui/ApplicationOverlay.cpp:
```diff
@@ -185,680 +90,75 @@ ApplicationOverlay::~ApplicationOverlay() {
+        geometryCache->useSimpleDrawPipeline(batch);
+        batch._glDisable(GL_DEPTH_TEST);
+        batch._glDisable(GL_LIGHTING);
+        batch._glEnable(GL_BLEND);
```
Not needed anymore, in the gpu::State for your pipeline you can set BLendFunc on etc

jherico replied:
fixed

samcake commented on interface/src/ui/ApplicationOverlay.cpp:
```diff
@@ -185,680 +90,75 @@ ApplicationOverlay::~ApplicationOverlay() {
+        batch._glEnable(GL_BLEND);
+        batch.setProjectionTransform(mat4());
+        batch.setModelTransform(mat4());
+        batch._glBindTexture(GL_TEXTURE_2D, _uiTexture);
```
ONce you have the _uiTexture as a gpu::Texture, just call Batch.setUniformTexture(mapSlot, _uiTexture)
mapSlot would be the slot where the shader expect to find that texture.
You can query that unit # from the Pipeline->ShaderProgram->getTextures().findLocation("THeNameOfTheTextureInTHeShader")

jherico replied:
the UI texture is generated by QML by another context.  We're never going to get it as a gpu::Texture.  If there is a mechanism to wrap a GL texture in a gpu::Texture, please let me know.

samcake commented on interface/src/ui/ApplicationOverlay.cpp:
```diff
@@ -990,40 +291,99 @@ void ApplicationOverlay::renderAudioMeter() {
+        gpu::Batch batch;
+        auto geometryCache = DependencyManager::get<GeometryCache>();
+        geometryCache->useSimpleDrawPipeline(batch);
+        batch._glDisable(GL_DEPTH_TEST);
```
Same comments as above regarding th euse of the _glCalls

hifi-gustavo commented:
Build console output for commit: 9bef5f011bc1da1e06aadd1d2c7f09a48fc187e8 available at: https://gist.github.com/9e3764e8f52d18446ef6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg

hifi-gustavo commented:
Build console output for commit: 99c217184b20833933273855cf2135a2ef753962 available at: https://gist.github.com/9e3764e8f52d18446ef6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg

hifi-gustavo commented:
Build console output for commit: cc6b589ae50fef8b292cfcf475be1755a975895a available at: https://gist.github.com/9e3764e8f52d18446ef6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg

hifi-gustavo commented:
Build console output for commit: 3ab8025b821afc46f5136914d33a2618da384e11 available at: https://gist.github.com/9e3764e8f52d18446ef6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg

PhilipRosedale commented:
Stuff I saw on testing (can't remember did I add in separate thread):  

1.  The skyboxes render OK (MBP on Desktop), but starfield is compressed into an oval.

2.  The text in the address bar is still blurry/aliased.

hifi-gustavo commented:
Build console output for commit: 577962cb6785eb0c7bcc45931fa3fc040fcc1aaf available at: https://gist.github.com/9e3764e8f52d18446ef6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg

hifi-gustavo commented:
Build console output for commit: e66b2d4141a5cd722dcc178585ac8b9e143cc1b1 available at: https://gist.github.com/9e3764e8f52d18446ef6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg

hifi-gustavo commented:
Build console output for commit: 5bbfb97c37351cc3259a418856b386e8aa7b3525 available at: https://gist.github.com/9e3764e8f52d18446ef6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg

jherico commented:
@PhilipRosedale found the source of the blurry UI, fixed.  Looking at the starfield UI issue now, however, I believe the starfield code should be ripped out and replaced with a shader that does the same thing.  

I have a placeholder starfield that doesn't produce the same distribution and color of stars here:  https://github.com/highfidelity/hifi/compare/highfidelity:master...jherico:starfield2

hifi-gustavo commented:
Build console output for commit: 012b0973b980c186cc4352da9050c3ffa9638ba5 available at: https://gist.github.com/9e3764e8f52d18446ef6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg

hifi-gustavo commented:
Build console output for commit: 4ccdc78866d61b398adc1360519dd5f1a0644ed3 available at: https://gist.github.com/9e3764e8f52d18446ef6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg

hifi-gustavo commented:
Build console output for commit: 02510790c38b66f3d4b6cf05f24da2118493f33f available at: https://gist.github.com/9e3764e8f52d18446ef6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg

jherico commented:
star rendering and text quality on overlays should both be fixed.

PhilipRosedale commented:
Yay!  Stars and text (both stats and address bar) look amazing now for me!

PhilipRosedale commented:
Looks good for me on my windows machine as well.  I'm not rendering my avatar in the mirror view and the world there is upside down, but that is true in the release version as well.

hifi-gustavo commented:
Build console output for commit: 1b44c220a729093141cb029ad0bb223b1af2dd03 available at: https://gist.github.com/9e3764e8f52d18446ef6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg

hifi-gustavo commented:
Build console output for commit: 3231bbba5ae78addda1f62bdf8d9f8f850d4983f available at: https://gist.github.com/9e3764e8f52d18446ef6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5125.dmg
